javalib/src/main/scala/java/util/PriorityQueue.scala

```diff
@@ -12,136 +12,181 @@
+  override def contains(o: Any): Boolean = {
+    if (o == null)
+      throw new NullPointerException()
+
+    val len = inner.length
+    var i = 1
+    while (i != len && !o.equals(inner(i))) {
+      i += 1
+    }
+    i != len
+  }
 
   def iterator(): Iterator[E] = {
     new Iterator[E] {
-      private val iter = inner.clone.iterator
-
-      private var last: Option[E] = None
+      private[this] var nextIdx: Int = 1
+      private[this] var lastIdx: Int = -1
 
-      def hasNext(): Boolean = iter.hasNext
+      def hasNext(): Boolean = nextIdx < inner.length
 
       def next(): E = {
-        last = Some(iter.next().inner)
-        last.get
+        if (!hasNext())
+          throw new NoSuchElementException("empty iterator")
+        lastIdx = nextIdx
+        nextIdx += 1
+        inner(lastIdx)
       }
 
       def remove(): Unit = {
-        if (last.isEmpty) {
+        if (lastIdx < 0)
           throw new IllegalStateException()
-        } else {
-          last.foreach(self.remove(_))
-          last = None
-        }
+        inner.splice(lastIdx, 1)
+        nextIdx = lastIdx
+        lastIdx = -1
       }
     }
   }
 
-  def size(): Int = inner.size
+  def size(): Int = inner.length - 1
 
   override def clear(): Unit =
-    inner.dequeueAll
-
-  def poll(): E =
-    if (inner.isEmpty) null.asInstanceOf[E]
-    else inner.dequeue().inner
-
-  def comparator(): Comparator[_ >: E] = _comparator
-}
+    inner.length = 1
+
+  def poll(): E = {
+    val inner = this.inner // local copy
+    if (inner.length > 1) {
+      val newSize = inner.length - 1
+      val result = inner(1)
+      inner(1) = inner(newSize)
+      inner.length = newSize
+      fixDown(1, newSize - 1)
+      result
+    } else {
+      null.asInstanceOf[E]
+    }
+  }
 
-object PriorityQueue {
+  def comparator(): Comparator[_ >: E] =
+    NaturalComparator.unselect(comp)
 
-  def safeGetOrdering[E](_comp: => Comparator[_]): Ordering[E] = {
-    val comp: Comparator[_] = _comp
-    val ord =
-      if (comp == null) defaultOrdering[E]
-      else Ordering.comparatorToOrdering(comp)
+  // Heavy lifting: heap fixup
 
-    ord.asInstanceOf[Ordering[E]]
+  private[this] def fixUp(m: Int): Unit = {
+    val inner = this.inner // local copy
+    var k: Int = m
+    while (k > 1 && comp.compare(inner(k >> 1), inner(k)) > 0) {
+      val j = k >> 1
+      val h = inner(k)
+      inner(k) = inner(j)
+      inner(j) = h
+      k = j
+    }
   }
 
+  private[this] def fixDown(m: Int, n: Int): Unit = {
+    // scalastyle:off return
+    val inner = this.inner // local copy
+    var k: Int = m
+    while (n >= 2 * k) {
+      var j = 2 * k
+      if (j < n && comp.compare(inner(j), inner(j + 1)) > 0)
+        j += 1
+      if (comp.compare(inner(k), inner(j)) <= 0) {
+        return
+      } else {
+        val h = inner(k)
+        inner(k) = inner(j)
+        inner(j) = h
+        k = j
+      }
+    }
+    // scalastyle:on return
```

Comment:
Same here: Why not write this tailrec? I think it will be much easier to understand.